Added collisions and scaling

theoden8 opened the pull request:
Hey! I have added collisions! Push my code to the master and we will remain friends !

theoden8 committed:
implemented collisions (in a quick and dirty way, wait until vectors arrive)

PolarBearSaviour commented on src/physical_object.py:
```diff
@@ -33,16 +33,39 @@ def get_forces_composition(self):
     def tick(self):
         """Tick the physical object: update its physical properties."""
+        s = self
         # print(str(self))
```
This can be removed

mallok-s commented on src/physical_object.py:
```diff
@@ -33,16 +33,39 @@ def get_forces_composition(self):
     def tick(self):
         """Tick the physical object: update its physical properties."""
+        s = self
         # print(str(self))
```
Can we remove commented out code please

theoden8 replied:
The reason to keep it there is to make debugging faster: we have all the debugging functions in correct places, whenever something goes wrong, we just uncomment them and faster navigate to the right place.

mallok-s commented on src/physical_object.py:
```diff
@@ -33,16 +33,39 @@ def get_forces_composition(self):
+
+        :returns: new speed
+        """
+        s = self
```
Shouldn't we being nouns as variable names? Might help improve the readability

theoden8 replied:
As long as it is clear what it takes, shouldn't matter. Although if I used p and q to represent different things in different places, that would make perfect sense.

mallok-s reviewed (commented):
The code seems quite clean. The s =self is a clever way of making the code shorter. Possible could use more detailed doc strings or comments. No idea if it works, so I don't think I can approve for merge (sorry, hope we'll remain friends!)

Commit:
Update physical_object.py

Removed comment out code... I didn't realise you could edit from the browser that's the only reason I commented on it

theoden8 committed:
more real collisions based on pair-pair colliding

theoden8 committed:
terminaet -> terminate

theoden8 committed:
improved collision

theoden8 committed:
added garbage (asteroids) collector and mass transfers after collision

theoden8 committed:
added docstrings for some functions related to collision

theoden8 commented:
@PolarBearSaviour, now it does work and very well. Please, reconsider: it is so lovely now! This is the feature to be proud of. 🎱 And yes, you are, of course, welcome to review it. Although it is monkey code, it is fast to write and easy to understand.

### Summary:

* The objects collide with full elasticity (complexity: N^2).
* A bit of mass of the smaller object transfers to the bigger object.
* When an object becomes very small, it is removed from the model as it is not displayable anyway, but causes laggs.
* Possible extension (doesnt work yet) of spawning asteroids after collision instead of transfering masses.

theoden8 committed:
updated min mass; added mass indicator to objects

theoden8 committed:
separated display coordinate system from physics engine; improved interface

theoden8 committed:
adjusted numbers: now it is somewhat a little bit more real. Yet raw and dirty

mallok-s reviewed (commented):
Kirill do you know if there issues with pygame and windows? It seems to work on mac (obviously) but on windows the code does nothing, like the screen just opens up and then nothing moves

theoden8 commented:
Interesting. PyGame is a python wrapper for SDL2, which is cross platform. If the computer is not very fast, increase the gravitational constant by a power of ten, or decrease the number of elements in the model. Either way, you should be able to see how fast does the word "tick" appear.

theoden8 committed:
made minor optimizations to speed up the program

theoden8 committed:
added a constant speedup: only recalculate forces every nth iteration

theoden8 committed:
fixed nth iterations bug: objects will now move

mallok-s reviewed (approved):
The code is fine and I can get it working on some windows machines. Potentially the code could contain more commentary but seems reasonable readable. Though I don't understand why we are using external libraries for physical constants, can we just ditch scripy and code in the constants ourselves?

theoden8 commented:
The reason to use external libraries for constants is because there might be something useful like light years, all in one. I just can't be bothered finding them on my own.
